spec/src/main/asciidoc/microprofile-lra-spec.adoc

```diff
@@ -816,17 +816,20 @@ exposed method from unauthorized access. The specific security details are not s
 ==== Eventual compensations
 
 [[async-compensators]]
-If the resource cannot perform a compensation activity
-immediately the `@Compensate` method SHOULD do one or more of the following:
+If a resource cannot perform a compensation activity
+immediately the `@Compensate` method MUST report that the activity
+is in progress as follows:
 
-- return `ParticipantStatus.Compensating`
-- return a `202 Accepted` HTTP status code
-- mark the method as a JAX-RS asynchronous method (using the
-  `javax.ws.rs.container.Suspended` annotation).
+- Return a CompletionStage or mark the method as asynchronous
+  (using the `javax.ws.rs.container.Suspended` annotation). The future must
+  deliver the final status when the stage completes. Please refer to the section
+  <<reactive-support,about reactive support>> for more details.
+- A JAX-RS method can return a `202 Accepted` HTTP status code
+- A non JAX-RS method can return `ParticipantStatus.Compensating`.
```

Comment:
This seems to be restricting usage of ParticipantStatus to non JAX-Rs methods? How can the participant indicate FailedToCompensate?

Reply:
The paragraph begins with:

> If a resource cannot perform a compensation activity immediately then ...

So if it wants to report FailedToCompensate then it means the activity has finished. In other  words this text only applies to compensation activities that are in progress.

Reply:
This text is in a section called "Eventual compensations" and the paragraph begins with the text:

>  If a resource cannot perform a compensation activity immediately then ...

it can return Compensating. If it has FailedToCompensate (one of the final states) then then this text is not applicable - ie the text only applies to compensations that are in progress.

Reply:
Exactly, so 

```java
return Response.ok(ParticipantStatus.FailedToCompensate).build()
```
 is allowed but 

```java
return Response.ok(ParticipantStatus.Compensating).build
``` 
isn't. I think this will be confusing for our users.

Reply:
LRA is JAX-RS based and therefore should conform to the expected REST norms and this will be particularly true when we add HTTP based interop in milestone 1.x.

When a participant is asked to compensate it should attempt to transition the resource into a final stated (Compensated or FailedToCompenstate). If the resource cannot make the transition immediately then the only reasonable REST based status code is 202:

> HTTP Status 202 indicates that request has been accepted for processing, but the processing has not been completed.

If the resource, or any intermediary, returns another response code then the caller cannot infer that the request is in progress.

Reply:
Why the caller cannot read the request body? It must read it for final states to distinguish between Compensated and FailedToCompensate. I understand your reasoning, but we should decide then if we want to follow the REST-based approach and allow JAX-RS participants to provide their state with HTTP statuses then we should follow it for everything. That means that HTTP OK "ParticipantStatus.FailedToCompensate" shouldn't be allowed as the participant has failed and some appropriate HTTP status (4xx/5xx) should be returned instead and it should not matter what the response body is (same as for 202 and Compensating case). This inconsistency as described in my previous comment will be confusing.

Reply:
I think the crux is whether FailedToComplete is a succesful outcome in our protocol: https://tools.ietf.org/html/rfc7231#section-6.3.1

As an aside, I definitely think a 4xx would be wrong as it would indicate a problem with the client call

Reply:
Also, reading 202 (https://tools.ietf.org/html/rfc7231#section-6.3.3) I think it should probably be 202 and a body containing `ParticipantStatus.Compensating` (or whatever)

Reply:
@tomjenkinson yes, you are right, sorry 5xx would be the right choice here.

My 2c: I would say all ParticipantStatus-es are valid outcomes because a participant is knowingly supplying the response. 200 is an indication of a successful REST request for compensate method, not to the LRA protocol.

Reply:
202 with a body of  `Compenstaing` to hold the status is fine (although superfluous).

Reaching an end state of `FailedToCompensate`, `Compensated` or `Completed` are all valid states for the resource to end with and therefore 200 is fine for those (and the caller can determine which state by reading the body).

As to responding with 200 for the state `Compensating`: the client is not asking the resource to move to some intermediate state, it wants the resource to move to a final state. If the request was "please start compensating" then 200 might be appropriate but the request is "please compensate" and responding with 200 would be wrong if it cannot immediately put the resource into a final state.

Reply:
> 202 with a body of `Compenstaing` to hold the status is fine (although superfluous).

Actually the body does make sense if the resource is already trying to complete when asked to compensate!

Reply:
> Why the caller cannot read the request body? It must read it for final states to distinguish between Compensated and FailedToCompensate. I understand your reasoning, but we should decide then if we want to follow the REST-based approach and allow JAX-RS participants to provide their state with HTTP statuses then we should follow it for everything. That means that HTTP OK "ParticipantStatus.FailedToCompensate" shouldn't be allowed as the participant has failed and some appropriate HTTP status (4xx/5xx) should be returned instead and it should not matter what the response body is (same as for 202 and Compensating case). This inconsistency as described in my previous comment will be confusing.

So if the outcome is anything other than `Compensated` or `Compensating` would returning `409 Conflict` with payload the actual status resolve your issue.

@tomjenkinson WDYT does section https://tools.ietf.org/html/rfc7231#section-6.5.8 apply here?

Reply:
So @xstefank,  for `@Compensate` notifications we would have:

> 200 if the status is `Compensated`
> 202 if the status is `Compensating`
> 409 if the status is `Active`, `Completed`, `FailedToCompensate` or `FailedToCompelete` (with the payload containing the actual status)

and similarly for response to the `@Complete` notification.

Reply:
> So if the outcome is anything other than `Compensated` or `Compensating` would returning `409 Conflict` with payload the actual status resolve your issue.

Not entirely. I am thinking we should either allow HTTP 200 with ParticipantStatus for all states OR don't allow it at all and allow only HTTP status codes for instance 200 == Compensated/Completed, 202 == Compensating/Completing, 5xx == FailedTo*. 

I actually don't see any issue with your original proposal from the sandbox where accepted is declared as 202 OR 200 with ParticipantStatus.Compensating/Completing OR @Suspended. This would only add 202 and @Suspended on top of what is already expected I think.

Reply:
So we have reached an impasse. `202 Accepted` should be the only valid response code to indicate that the request to compensate is in progress.

Reply:
I don't think `Active` is valid at all for the compensator asked to `@compensate` the rest of the codes I liked WRT https://tools.ietf.org/html/rfc7231#section-6.5.8

Reply:
2> So we have reached an impasse. `202 Accepted` should be the only valid response code to indicate that the request to compensate is in progress.

@rdebusscher @ochaloup WDYT?

Reply:
Not very familiar with asynchronous REST, but from what I know and read, when using @Suspended the method should return void.

The CompletionStage/Future is then used on the client side when such an async REST is called.

JAX-RS case
-------------

- return status 202 (if this is required and not handled out of the box by JAX_RS if we use @Suspended)
- AsyncResponse.resume(ParticipantStatus.Compensated) or AsyncResponse.resume(ParticipantStatus.FailedToCompensated) are allowed

non JAX-RS case
-----------------

- Method returns a CompletionStage
- returned value should be _ParticipantStatus.Compensated_ or _ParticipantStatus.FailedToCompensated_


The same logic applies for the Complete scenario with the appropriate ParticipantStatus.

Reply:
I don't think the `202 Accepted` is the only valid response. I can see the perspective  that the `Compensating` response could be rated as a success. This response is used by a user - it's not handled by framework - so from "API" design perspective it's good to take into account how consistent is the use of the return values. If the proposed rules are comprehensible.

Nevertheless, I think differentiating the response by the HTTP codes is what the REST api should strive for.

I understand `ParticipantStatus.FailedToCompensate` changes the the state to a final state but it's an error state. I don't think the `Response.ok(ParticipantStatus.FailedToCompensate).build()` is valid (if we talk about HTTP error codes to be used). If it's valid then the `Response.ok(ParticipantStatus.Compensating).build()` should be considered to be valid as well.
From that perspective using http status `409` sounds good to me.

Reply:
From what I am reading the views are:

- @tomjenkinson is okay with 202 for compensating, 200 for compensated and 409 for failure states. For terminal states the body contains the status. For non-terminal states we have a choice: the status URL can be placed in the body or location header (we had this option when we were in the sandbox) or we leave the body empty or we include the status..

- @rdebusscher is okay with 202 for `Compensating`. Use of `@Suspended` is not an issue since a synchronous client is blocked until the server async method completes (ie it will only return when a final state is reached). And an async client will gather the completed result when the server makes it available so again it should never see an intermediate state.

- @ochaloup I am not sure what you want, you seem to be arguing a case for all options, ie you are seeing pros and cons for each alternative.

- @xstefank is okay with 202 for `Compensating`, 200 for `Compensated` and 5xx for the failure states. He is also okay with 200 + the status in the body for all states. But he did add a third alternative of supporting two ways to report `Compensating` depending upon whether the body is present or not.

- @mmusgrov JAX-RS is for REST based applications and the recommendation for reporting asynchronous/in-progress operations is to use the code `202 Accepted`. The status code returned by a participant that has reached a final state is a matter of taste between 200 or 409. 409 means "The request could not be completed due to a conflict with the current state of the target resource" and that matches this situation (the client asks for state Compensated but the resource is in a state where that is not possible - ie there is a conflict between what the client is asking for and what the resource state is).

So from the above feedback the only set of codes that everyone has some degree of agreement with is:

> 202 for compensating, 200 for compensated and either 200 or 409 or 5xx for failure states

Can we agree on 200 or 409 for failure states (using my argument in the previous paragraph for 409). I do not see a case for 5xx codes (which equate to s/w or h/w or configuration failures).

Reply:
@mmusgrov ok, so let's go with 200 Compensated, 202 Compensating and 409/5xx(user exceptions) FailedToCompensate and ignore the body of the response. This is a clear distinction.

Reply:
Not sure about 409

>This code is used in situations where the user might be able to resolve the conflict and resubmit the request.

But if everyone else agree, I'm fine with it

Reply:
It doesn't matter to me what will be the error code but it just must be unique. @rdebusscher do you have something else in mind, please?

Reply:
@mmusgrov yes, I was not able to clearly decide as the cons and pros are quite in balance.
Distinction based on the error codes sounds fine to me.

Reply:
200 = Reached the expected state
202 = In progress
But as to whether it is 200 or 500 (not 5xx, looking at the list of these the only one I could argue for would be [500](https://tools.ietf.org/html/rfc2616#section-10.5.1)) for FailedTo<> depends on your point of view but I think 500

Reply:
I was thinking that 200 would be correct as we do ask the participant to `compensate` and valid outcome states are either `compensated` or `failedtocompensate` and these should be stable outcomes until `forget` is called IIUC.

However then I think I found the definitive position that it should be 500 due to the wording "any other exception" in 3.2.5 in on https://github.com/eclipse/microprofile-lra/blob/master/spec/src/main/asciidoc/microprofile-lra-spec.adoc.

Therefore, I wonder if we can agree on 200, 202 and 500?

Reply:
> Not sure about 409
> 
> > This code is used in situations where the user might be able to resolve the conflict and resubmit the request.
> 
> But if everyone else agree, I'm fine with it

It does say "might". It also says the payload SHOULD tell the caller what to do.

In our case the coordinator would read the payload, discover the reason for the conflict, and see that there is no point in resubmitting the request.

I am still okay with 200 or 409 for the failure states.

Reply:
The IETF description of 409 does have the nice feature of saying the server SHOULD include enough information to enable the caller to recognise the reason for the conflict. And our spec, as it stands, already provides this information by returning `FailedToCompensate` in the payload.

But if people think 409 is wrong then I am inclined to agree with Tom even though the section he cites refers to non JAX-RS compensation methods so JAX-RS codes are not so applicable. 500 does not provide any reason for the failure and expecting a semantically meaningful payload is probably wrong. Which brings me back to 200. This is fun.

Reply:
I was thinking that the reason it was similar is the non-JAX-RS version reports FailedToComplete as an _exception_ rather than a valid outcome:

In other words rather than having something like:

```public ParticipantCompensateStatus compensate(....) {}```

We went for: 

```public void compensate() {
// If I can't do this
throw new RuntimeException("I failed to compensate");
}```

I did like your point about [409](https://tools.ietf.org/html/rfc2616#section-10.4.10) and returning enough information to recongize the source of the conflict but then after reading some more I am not sure if we would be fulfilling this part:
> "This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request" 
as I don't know that the user (LRA coordinator) can resolve the conflict and resubmit the request unless we think that by calling @Forget on the resource and then calling @Compensate and getting a 404 (which allows the user - LRA coordinator - know things are OK) is an acceptable way to "resolve the conflict" - perhaps it is...

Reply:
> I did like your point about [409](https://tools.ietf.org/html/rfc2616#section-10.4.10) and returning enough information to recongize the source of the conflict but then after reading some more I am not sure if we would be fulfilling this part:
> 
> > "This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request"
> 
> as I don't know that the user (LRA coordinator) can resolve the conflict and resubmit the request unless we think that by calling @forget on the resource and then calling @compensate and getting a 404 (which allows the user - LRA coordinator - know things are OK) is an acceptable way to "resolve the conflict" - perhaps it is...

I addressed that concern in my response to Rudy, [see comment](https://github.com/eclipse/microprofile-lra/pull/176#discussion_r301199890)

Reply:
But I don't think we will be able to reach any consensus on this so would taking a poll help to make progress. The options for reporting a failure to compensate are as follows:

1. 200 + body of `FailedToCompensate`
2. 409 + body of `FailedToCompensate`
3. 500 + body of `FailedToCompensate`
4. ??? (if you want to vote for something else)

The requirement for a payload in option 3 is to handle the case where the response was returned by an intermediary.

If the body does not indicate the reason then the implementation MUST retry the request or use the `Status` URI.

Reply:
And if voting is a viable resolution please add a +1 or -1 for each option so that we can pick a winner.

Reply:
> 
> 
> > I did like your point about [409](https://tools.ietf.org/html/rfc2616#section-10.4.10) and returning enough information to recongize the source of the conflict but then after reading some more I am not sure if we would be fulfilling this part:
> > > "This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request"
> > 
> > 
> > as I don't know that the user (LRA coordinator) can resolve the conflict and resubmit the request unless we think that by calling @forget on the resource and then calling @compensate and getting a 404 (which allows the user - LRA coordinator - know things are OK) is an acceptable way to "resolve the conflict" - perhaps it is...
> 
> I addressed that concern in my response to Rudy, [see this comment](https://github.com/eclipse/microprofile-lra/pull/176#discussion_r301199890)

That comment didn't address the "only allowed" element, just the "might" and "resubmit the request" was not addressed either. But I think my answer proposed sequence does address it more concretely as they client can resubmit (after calling forget), obtain the 404 and treat that differently.

Anyway, I do agree it is time for a poll.

I vote:

1. +0
2. +1
3. +1

Reply:
/cc @rdebusscher @xstefank @ochaloup please can you all provide a vote on the four options?

Reply:
> 
> 
> And if voting is a viable resolution please add a +1 (yes) or -1 (no) or 0 (abstain) for each option so that we can pick a winner. So my votes are:
> 
>     1. +1
> 
>     2. +1
> 
>     3. 0

I forget:
4. 0

Reply:
1. -1
2. +1
3. +1

Reply:
1. +1 (when developer explicitly wants to set the FailedToCompensate status)
2. -1
3. +1 (When it failed due to an error)
4. -1

Reply:
1. -1
2. -1
3. +1

Please let's use either HTTP codes **or** ParticipantStatus response body. If we are going to mix them, we should also allow 200 Compensating which was the original issue. So I don't really have an issue with whether it is 409 or 500, but it must be only one of them and the body is ignored. @rdebusscher developer can also explicitly set 500 HTTP status code, wouldn't that be the same as setting FailedToCompensate response body?

Reply:
@xstefank Yes, but I prefer

> public ParticipantStatus doCompensate(...)

over

> public Response doCompensate(...)

or

> throw new WebApplicationException(500)

Reply:
I was thinking we are talking only JAX-RS participants here.

Reply:
I would say option 3 "500 + body of FailedToCompensate" is the winner. I reached that conclusion as follows:

Counting votes gives the result:

1. 0
2. 3
3. 4
4. -1

We nearly got an unambiguous winner with option 3: "500 + body of FailedToCompensate" but Martin has muddied the waters somewhat by voting for a fourth option since he added a caveat after his vote which means he is actually voting for -1 on questions 1, 2 and 3 and +1 on option 4 which would produce the following results:

1. 0
2. 2
3. 3
4. 0

Even with this interpretation of the voting, option 3 is still the winner - can we agree to go with that then?

Reply:
@xstefank 

> The requirement for a payload in option 3 is to handle the case where the response was returned by an intermediary

ie voting for option 3 requires a payload. Also this vote is only for reporting failure states.

All of the other states can be reported without a payload.

Reply:
@mmusgrov sorry for the misunderstanding. I am standing by 500 even if I don't agree with the necessity of payload. But I was overvoted so let's see how this plays out.

Reply:
I am interested to know why you think a payload isn't required for the 500 case since if we can avoid it then the requirements on JAX-RS participants would be much cleaner.

But if we do agree on option 3 then when a JAX-RS participant is asked to compensate it MUST respond, depending upon its state, with:

State  HTTP Code Payload
-----    ---------------- -------

Compensated 200 Empty
Compensating 202 Empty
FailedToCompensate 500 "FailedToCompensate"

(400 is valid if the participant does not exist on the target server).
In the failure case, if the body is not FailedToCompensate the implementation MUST retry until it has the final state.
Once it has the final failure state it MUST call `Forget` on the participant.

Reply:
> In the failure case, if the body is not FailedToCompensate the implementation MUST retry until it has the final state.

What about `@Status` method? Is it respected in this case?

Reply:
Question: In the FailedToCompensate state do we want to include in the payload an indication of why it failed (using one of the ParticipantStatus enum values: Completed, Completing, FailedToComplete). Is that something that would ever be of interest? It could be used to enhance the `@AfterLRA` notification in milestone 1.x

Reply:
> > In the failure case, if the body is not FailedToCompensate the implementation MUST retry until it has the final state.
> 
> What about `@Status` method? Is it respected in this case?

Yes it is, sorry for the ambiguity. I will make that clear when I update the PR.

But can you indicate how we can avoid a payload when the participant is reporting `FailedToCompensate`. I am concerned that the HTTP server could return 500 even though the participant is just fine.

Reply:
Personally, I would prefer to stay with just payload of ParticipantStatus values and 200 HTTP status code. But since we've already decided that we will map the LRA protocol on top of HTTP status codes then we should have enough information in the status meaning - 200 Compensated, 202 Compensating, 500 FailedToCompensate. For 500 it doesn't matter whether the user explicitly passed 500: `return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();` or if some error was thrown in the compensate method and framework returned 500 (or similar).

I can see where we got confused, I am not expecting that we will call the compensate again when there is failure on the participant side (5xx). In that case, some distinction is necessary.

Reply:
I would be concerned about relying purely on status code 500. Is it possible that the /compensate method executes fine and returns a 200 but some intermediary server (proxy?) fails and returns the 500?

Reply:
That is true @tomjenkinson. In that case wouldn't the response body be overridden with the new error?

Reply:
I thought we were investigating ways to avoid having to look at the response body in the 500 case. If we could do that then that would be great but there is also the case where the request is intercepted by some processing entity before it reaches the participant. A JAX-RS server for example has an internal error that means it cannot deliver the message to the resource and returns 500. In this case the participant is still in the active state (not FailedToCompensate).

Reply:
That would be my concern. I think we have to check it is 500 and the body contains specific content to prevent the situation. I still think it is valid to consider 500 as a reasonable response code even if you need to look in the body on the client side given the definition of 5xx includes "Except when responding to a HEAD request, the
   server SHOULD include an entity containing an explanation of the
   error situation, and whether it is a temporary or permanent
   condition."

Reply:
When we have resolved this current thread I will prepare an update to the PR [according to the voting results.](https://github.com/eclipse/microprofile-lra/pull/176#discussion_r301567046)

Reply:
I am acking the results are OK with me

Reply:
@mmusgrov agreed. If we are moving to HTTP status codes, we will have to have the response body for 500. 

So the final outcome would be:

200 - Compensated/Completed
202 - Compensating/Completing
500 + FailedToCompensate, 500 + FailedToComplete - FailedToCompensate/FailedToComplete
500 (no response  body) - Compensating/Completing

------------------------------------------------------------------------------------------------------------------------
This is just to document my view, feel free to ignore.

I still think that stating participant status through HTTP response code is not a good way of handling this because:
* mainly: LRA is a protocol layer on top of HTTP/REST. HTTP/REST is for us only the transport protocol.
* this effectively cannot be reused if we ever introduce different transport protocols 
* it introduces the possibility that users will forget about including response bodies (in 500 cases) which will result in unexpected behavior
* it forces users to use javax.ws.rs.Response object as return type so something like this is no longer possible: 
```
    @PUT
    @Path("/compensate")
    @Produces("text/plain")
    @Compensate
    public ParticipantStatus compensate(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
        return isOk(lraId) ? ParticipantStatus.Compensated : ParticipantStatus.FailedToCompensate;
    }
```
-------------------------------------------------------------------------------------------------------------------------

so +1 on moving this forward with the decided solution

Reply:
I think the reason we can't support that code is we got -1 from two voters on allowing 200 for FailedToCompensate? But maybe I misunderstood?

Reply:
> I think the reason we can't support that code is we got -1 from two voters on allowing 200 for FailedToCompensate? But maybe I misunderstood?

@tomjenkinson please ignore that, I just wanted to document why I think it would be better to stick with 200. But please ignore that, I am +1 on the voted solution.

Reply:
@mmusgrov this can be resolved from my side. Please update the PR with the voted results.

Reply:
> * mainly: LRA is a protocol layer on top of HTTP/REST. HTTP/REST is for us only the transport protocol.

- The LRA spec is for JAX-RS resources to be able to participate in LRAs and JAX-RS is for REST. Therefore using JAX-RS response codes should be the norm
- We would like to use HTTP for interop between implementations

> * this effectively cannot be reused if we ever introduce different transport protocols

It's a good point but the spec is JAX-RS specific

> * it introduces the possibility that users will forget about including response bodies (in 500 cases) which will result in unexpected behavior

True. If there is status method that will remove any ambiguity and if that isn't available then all I can suggest is that the implementation logs something or the developer has a test suite to catch the problem.

> * it forces users to use javax.ws.rs.Response object as return type so something like this is no longer possible:
> 
> ```
>     @PUT
>     @Path("/compensate")
>     @Produces("text/plain")
>     @Compensate
>     public ParticipantStatus compensate(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
>         return isOk(lraId) ? ParticipantStatus.Compensated : ParticipantStatus.FailedToCompensate;
>     }
> ```
> 

Yes that is a shame since it forces the developer to use a particular coding style. On the plus side using a javax.ws.rs.Response should be natural for JAX-RS developers.

> so +1 on moving this forward with the decided solution